### PR TITLE
W-12356140: [SingleApp] Avoid on-demand artifact deployment

### PR DESCRIPTION
**Pull request by alopezbizon**

### CODE REVIEW

**Comment by alopezbizon** [comment]
--validate

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/ManagerDeploymentDirectoryWatcher.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages the deployment directory needed.
+ *
+ * @since 4.7.0
+ */
+public class ManagerDeploymentDirectoryWatcher {
```
Comment (fsgonz) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/ManagerDeploymentDirectoryWatcher.java:37]: This should be named as a provider or something like that. This is not a watcher but obtains the watcher.

**Comment by alopezbizon** [comment]
--validate

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2023 Salesforce, Inc. All rights reserved.
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.deployment.internal;
+
+import static java.lang.System.getProperties;
```
Comment (fsgonz) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java:9]: static import order.

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  public DeploymentExecutor(DeploymentService deploymentService) {
+    this.deploymentService = deploymentService;
+  }
+
+  public <D extends DeployableArtifactDescriptor, T extends Artifact<D>> void deploy(final URI artifactArchiveUri,
+                                                                                     final Optional<Properties> deploymentProperties,
+                                                                                     File artifactDeploymentFolder,
+                                                                                     ArchiveDeployer<D, T> archiveDeployer)
+      throws IOException {
+    if (!isSingleDeployment() || (isSingleDeployment() && !this.artifactDeployed)) {
```
Comment (fsgonz) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java:40]: try to create a polymorphism where this "if" is taken into account. An interface and two classes and create the deployment service accordingly in the mule container as talked.

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+public class DeploymentExecutor {
+
+  private boolean artifactDeployed = false;
+  private final DeploymentService deploymentService;
+
+  public DeploymentExecutor(DeploymentService deploymentService) {
+    this.deploymentService = deploymentService;
+  }
+
+  public <D extends DeployableArtifactDescriptor, T extends Artifact<D>> void deploy(final URI artifactArchiveUri,
```
Comment (fsgonz) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java:35]: jdoc.

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+          archiveDeployer.deployExplodedArtifact(fileName, deploymentProperties);
+        }
+      } catch (MalformedURLException e) {
+        throw new MuleRuntimeException(e);
+      }
+    });
+  }
+
+  private boolean isSingleDeployment() {
+    return getProperties().containsKey(SINGLE_DEPLOYMENT_PROPERTY);
```
Comment (fsgonz) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java:74]: Evaluate this once and not every time you invoke the method. The sysprops will not vary.

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    this.deploymentService = deploymentService;
+  }
+
+  public <D extends DeployableArtifactDescriptor, T extends Artifact<D>> void deploy(final URI artifactArchiveUri,
+                                                                                     final Optional<Properties> deploymentProperties,
+                                                                                     File artifactDeploymentFolder,
+                                                                                     ArchiveDeployer<D, T> archiveDeployer)
+      throws IOException {
+    if (!isSingleDeployment() || (isSingleDeployment() && !this.artifactDeployed)) {
+      this.artifactDeployed = true;
```
Comment (fsgonz) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java:41]: verify that this is set as deployed after you know that you really deployed the artifact, not when you attempt to deploy it.

**Review by fsgonz — CHANGES_REQUESTED** [review]

**Review by fsgonz — APPROVED** [review]

**Comment by alopezbizon** [comment]
--validate

**Pull request merged**